### ISSUE TITLE
reserved identifier violation

### ISSUE THREAD

**Issue by elfring** [01org/tbb#75]
I would like to point out that identifiers like “[`__itt_caller`](https://github.com/01org/tbb/blob/615bb971db8784ab6c005a856e8ed3e8cea61d9b/include/tbb/task.h#L29 "Update candidate")” and “[`__TBB_exception_H`](https://github.com/01org/tbb/blob/d31fcc68980efbe912dc8cec4978fbb77fb3cf9f/include/tbb/tbb_exception.h#L21 "Another update candidate")” [do eventually not fit](https://www.securecoding.cert.org/confluence/display/cplusplus/DCL51-CPP.+Do+not+declare+or+define+a+reserved+identifier "Do not declare an identifier which is reserved for the compiler implementation.") to the expected naming convention of the C++ language standard.
Would you like to adjust your selection for unique names?

**Comment by alexey-katranov** [comment]
@elfring, names starting with a double underscore `__` really reserved to the implementation and usually should be avoided. However, we use such names to avoid issues with names used by our customers. We restrict all our names to start with `__TBB*` and our tests show that our approach does not conflict with existing C++ implementations. So if our customers follow [the mentioned rule](https://www.securecoding.cert.org/confluence/display/cplusplus/DCL51-CPP.+Do+not+declare+or+define+a+reserved+identifier), the approach eliminates name conflict risks.

**Comment by elfring** [comment]
How do you think about to avoid that this software [depends on undefined behaviour](https://www.securecoding.cert.org/confluence/display/c/CC.+Undefined+Behavior#CC.UndefinedBehavior-ub_106 "Avoiding undefined behaviour?")?

**Comment by alexey-katranov** [comment]
Strictly speaking, you are right that using names in implementation reserved domain is undefined behavior. However, let's consider the conflict risk problem for the whole application that uses C++ implementation, TBB library and user application code. The name conflicts usually lead to misleading compilation issues (in some extraordinary cases it can lead to unexpected behavior in runtime). To mitigate the risks of name conflicts, we may use one of the following approaches:

- Use fully standard complaint names, document all used names and enforce our users to follow standard rules and our library rules;
- Use names in implementation reserved domain and test that they are compatible with supported C++ implementations.

We believe that the second approach is more convenient (or even safer) for our customers for majority of applications. Because they only need to follow standard naming rules and can consider our library as a part of the system.

**Comment by alexey-katranov** [comment]
> 🔮 How strict would you like to handle the “conflict risk problem”?

Difficult question... In majority of cases a name conflict will be detected at compile time easily. However, if some data types are redefined (e.g. `BOOL` in `C` can be `int` but in `C++` it can be `bool`), it might not be detected at compile time and might lead to serious memory violation issues.

> Software designers and developers chose identifiers including from the reserved namespace for this library for a while.

I suppose, it has become a common practice (not only in this library but in the whole industry) to use `*_t` while it is not compatible with POSIX. Perhaps, this common practice should be challenged.

> I would interpret such a design choice in the way that this library can be safely only used for the implementation of compilers for the programming languages “C” and “C++” so far.

I am not sure that understand the idea. The TBB library is written in C++ and can be used only with C++ compiler.

> Your software users (or involved companies) might not care so much about this aspect of standard compliance.

If the customers does not care about this aspect of the standard, I suppose they will not care about naming aspects of some third-party library. That is one of the motivation to use reserved names to reduce the scope of possible issues for the customers.

> How are the chances to achieve a renaming of affected identifiers so that this library can be safely used also for all applications (without restrictions)?

Technically, it does not matter how identifiers are named. However, hypothetical renaming might extend the list of reserved identifiers for our customers, i.e. they should care and C++ naming, POSIX naming and third-party libraries naming rules. From safety perspective, it might be the right approach but it might require customers to spend more efforts (resources) to follow (and, perhaps, lead to some violations because they need to follow more rules than previously).

**Comment by elfring** [comment]
>I suppose, it has become a common practice (not only in this library but in the whole industry) to use *_t

Will it become interesting to look more around the attractiveness for such a suffix?


>while it is not compatible with POSIX.

It can be compatible [according to specific constraints](https://en.cppreference.com/w/c/language/identifier#Reserved_identifiers "Description for reserved identifiers").


>Perhaps, this common practice should be challenged.

The applied software design preferences can be reconsidered, can't they?


>I am not sure that understand the idea. The TBB library is written in C++ and can be used only with C++ compiler.

The standard language specifications express that reserved identifiers may be used by “the implementation”.
* Which software components will be involved then there?
* Was any compiler implemented with the help of this library?
* It will be nice if such a library can be safely used also for other kinds of applications, won't it?


>…, I suppose they will not care about naming aspects of some third-party library.

I guess that this is matter of awareness and the usual distribution of possible software development attention.


>That is one of the motivation to use reserved names to reduce the scope of possible issues for the customers.

There are further software design possibilities to consider besides tampering with the reserved name space.


>From safety perspective, it might be the right approach

Thanks that you can take the renaming of affected identifiers into account.

Do know any special tools which can make a corresponding source code transformation approach more convenient and safe finally?


>but it might require customers to spend more efforts (resources) to follow …

I get the impression that this view can distract from desired solutions.
:crystal_ball: Should all needed software libraries be implemented for unrestricted usage in various applications?

**Comment by elfring** [comment]
I would appreciate another constructive feedback for this issue.

**Comment by akukanov** [comment]
@elfring, this issue is closed because it is highly unlikely that the current practice will change.

We admit that use of leading underscores results in undefined behavior and poses certain risks, primarily for code portability. But we find these risks very low in practice, which is confirmed by 12 years of TBB use in the field, and by a number of successful ports to environments that were not originally supported. Note that we do not use underscores blindly but follow certain rules to further minimize the risk of conflicts with names used by implementations. I cannot recall a single case where our names starting with underscores conflicted with any ones used by a C++ implementation. Also, as Alex mentioned, it is not obvious whether the risk of name conflicts would be reduced or increased for programs that use TBB.

The bottom line is that this is considered a low priority issue with low (if any) ROI.

For a general discussion about acceptability of such naming approach, it is better to use some public forum or mailing list. We will not continue the discussion here.

**Comment by elfring** [comment]
I find your handling of undefined behaviour questionable.
Will any programming standards require to increase your software development attention also for such aspects?
